### ISSUE TITLE
label nodes with the name of the autoscaling group they belong to (if they belong to one)

### ISSUE THREAD

**Issue by tkellen** [kubernetes/cloud-provider-aws#884]
**What would you like to be added**:
It would be extremely helpful for operators if nodes were labelled with the autoscaling group they belong to (if there is an autoscaling group associated with a given node). I think `node.kubernetes.io/auto-scaling-group-name` would be appropriate for this.

**Why is this needed**:
Alleviate the need to cross-reference autoscaling group instances in AWS with node names in kubectl.

I would be happy to submit a PR if this would be considered for acceptance. Let me know!

/kind feature

**Comment by mmerkes** [comment]
I don't think this is unreasonable as long as we don't need to add any additional API calls, and it looks like `aws:autoscaling:groupName` is set on the instances by ASG, so we can just read the tag and apply the label with no additional calls. Thoughts @cartermckinnon?

**Comment by cartermckinnon** [comment]
Sounds reasonable, looks like we'll get the tags from `ec2:DescribeInstances`. ASG guarantees that will be added at creation, not after launch: https://docs.aws.amazon.com/autoscaling/ec2/userguide/ec2-auto-scaling-tagging.html#tag-lifecycle

Took a look at the code paths and it's a bit of a mess. We need to reuse the `ec2:DescribeInstances` response in our `InstancesV2.InstanceMetadata` implementation, we have a few redundant calls today. But we can pass this with the new `AdditionalLabels` field.

I think we'll want to use a label under the `*.k8s.aws` namespace, not `node.kubernetes.io`

**Comment by tkellen** [comment]
🙏🏻🙏🏻 the complexity reduction this would bring to my k8s node group upgrade script would be massive 🤞🏻.

**Comment by mmerkes** [comment]
+1 to what Carter is saying. [There's currently one well known label in the cloud provider](https://github.com/kubernetes/cloud-provider-aws/blob/master/pkg/providers/v1/well_known_labels.go#L22): `topology.k8s.aws/zone-id`. Something like `node.k8s.aws/auto-scaling-group-name` might be the one.

**Comment by tkellen** [comment]
Is this something you're looking for an outside contributor to implement (not yet sure how I would test it or I would have made an attempt already) or should I sit tight and let y'all do your thing?

**Comment by cartermckinnon** [comment]
I'll put something together, I'd like to do some cleanup anyway 😄

**Comment by tkellen** [comment]
I am seeing the following labels applied to new nodes:
`eks.amazonaws.com/nodegroup`
`eks.amazonaws.com/sourceLaunchTemplateId`
`eks.amazonaws.com/sourceLaunchTemplateVersion`

Out of curiosity, where was this implemented? I don't see it in this repo?

**Comment by mmerkes** [comment]
> I am seeing the following labels applied to new nodes: `eks.amazonaws.com/nodegroup` `eks.amazonaws.com/sourceLaunchTemplateId` `eks.amazonaws.com/sourceLaunchTemplateVersion`
> 
> Out of curiosity, where was this implemented? I don't see it in this repo?

That's not part of the cloud provider. You must be using EKS Managed Node Groups, which applies those labels.

**Comment by tkellen** [comment]
Ah! That makes sense, you're absolutely right. So many clusters. So much config 😵‍💫.

**Comment by tkellen** [comment]
/remove-lifecycle-rotten

checking in again to see if y'all are desirous of a contribution for this? some small guidance on testing and the cleanup you'd hoped to implement would be enough for me to dig in, I imagine.

**Comment by tkellen** [comment]
Anyone out there to respond to https://github.com/kubernetes/cloud-provider-aws/issues/884#issuecomment-2290527911? I would be glad to implement this.

**Comment by mmerkes** [comment]
> Is this something you're looking for an outside contributor to implement (not yet sure how I would test it or I would have made an attempt already) or should I sit tight and let y'all do your thing?

This project is maintained by the open source community and contributions are always welcome, so feel free! If the implementation isn't overly complex, shouldn't be hard to get someone to take a look.

As for testing, I don't know the best way to get a local setup going. However, with proper unit tests, you might get enough confidence in your change to put up a PR with an integration test that the CI can run and validate.

**Comment by lmktfy** [comment]
> I think `node.kubernetes.io/auto-scaling-group-name` would be appropriate for this.

How about using an AWS domain name for the first part of the label key? We are specifically talky about the AWS concept of an autoscaling group.

**Comment by luckyganesh** [comment]
> +1 to what Carter is saying. [There's currently one well known label in the cloud provider](https://github.com/kubernetes/cloud-provider-aws/blob/master/pkg/providers/v1/well_known_labels.go#L22): `topology.k8s.aws/zone-id`. Something like `node.k8s.aws/auto-scaling-group-name` might be the one.

@lmktfy this looks good to me, open for suggestions

**Comment by lmktfy** [comment]
If AWS sign off - see https://github.com/kubernetes/cloud-provider-aws/pull/1239#discussion_r2543918858v - then great. If not, we could get a new domain name and use that!
